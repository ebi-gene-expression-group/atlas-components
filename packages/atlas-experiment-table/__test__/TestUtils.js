import React from 'react'
import styled from 'styled-components'

// Stolen from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomInt = (min, max) => {
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(Math.random() * (max - min)) + min //The maximum is exclusive and the minimum is inclusive
}

const TableCellDiv = styled.div`
  font-size: 13px;
  font-family: Helvetica, Arial, FreeSans, "Liberation Sans", sans-serif;
`

const tableHeader = [
  { type: `sort`, title: `Loaded date`, width: 140, dataParam: `lastUpdate` },
  { type: `search`, title: `species`, width: 200, dataParam: `species` },
  { type: ``, title: `experiment description`, width: 360, dataParam: `experimentDescription`, link: `experimentAccession`, resource: `experiments`, endpoint: `Results` },
  { type: `plain`, title: `experiment factors`, width: 260, dataParam: `experimentalFactors` },
  { type: `sort`, title: `Number of assays`, width: 160, dataParam: `numberOfAssays`, link: `experimentAccession`, resource: `experiments`, endpoint: `Experiment Design` },
]

const data = [
  {
    experimentType: `SINGLE`,
    experimentAccession: `E-EHCA-2`,
    experimentDescription: `Melanoma infiltration`,
    lastUpdate: `16-11-2018`,
    numberOfAssays: 6638,
    numberOfContrasts: 0,
    species: `Mus musculus`,
    kingdom: `animals`,
    experimentalFactors: [`single cell identifier`, `sampling site`,`time`],
  },
  {
    experimentType: `SINGLE`,
    experimentAccession: `E-GEOD-99058`,
    experimentDescription: `Single cell`,
    lastUpdate: `11-10-2018`,
    numberOfAssays: 250,
    numberOfContrasts: 0,
    species: `Mus musculus`,
    kingdom: `animals`,
    experimentalFactors: [`single cell identifier`]
  },
  {
    experimentType: `SINGLE`,
    experimentAccession: `E-MTAB-5061`,
    experimentDescription: `healthy individuals and type 2 diabetes patients`,
    lastUpdate: `11-10-2018`,
    numberOfAssays: 3514,
    numberOfContrasts: 0,
    species: `Homo sapiens`,
    kingdom: `plants`,
    experimentalFactors: [`single cell identifier`, `disease`],
  }
]

export {getRandomInt, TableCellDiv, tableHeader, data}
