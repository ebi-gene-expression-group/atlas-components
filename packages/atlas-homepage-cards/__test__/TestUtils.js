

// Stolen from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomInt = (min, max) => {
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(Math.random() * (max - min)) + min //The maximum is exclusive and the minimum is inclusive
}

const dummyCards = () => {
  return [
    {
      iconType: `species`,
      iconSrc: `mus musculus`,
      iconDescription: `Mus musculus`,
      content: [
        {
          text: `4 experiments`,
        }
      ]
    },
    {
      iconType: `species`,
      iconSrc: `homo sapiens`,
      iconDescription: `Homo sapiens`,
      content: [
        {
          text: `2 experiments`,
        }
      ]
    },
    {
      iconType: `unknown`,
      iconSrc: `http://www.lingyun.com`,
      iconDescription: `Mus musculus`,
      content: [
        {
          text: `4 experiments`,
        }
      ]
    },
    {
      iconType: `experiments`,
      iconSrc: `http://www.zhao.com`,
      iconDescription: `Homo sapiens`,
      content: [
        {
          text: `2 experiments`,
        }
      ]
    },
    {
      iconType: `image`,
      iconSrc: `logo-goes-here`,
      iconDescription: `An image card`,
      content: [
        {
          text: `Go to experiments`,
        }
      ]
    },
    {
      iconType: `imagespecies`,
      iconSrc: `http://www.lingyun.com`,
      iconDescription: `Mus musculus`,
      content: [
        {
          text: `4 experiments`,
        }
      ]
    },
  ]
}

export {getRandomInt, dummyCards}