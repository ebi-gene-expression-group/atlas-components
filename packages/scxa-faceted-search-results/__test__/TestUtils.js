import React from 'react'
import PropTypes from 'prop-types'

const EpisodeCard = ({title}) =>
  <div>
    <p>{title}</p>
  </div>

EpisodeCard.propTypes = {
  title: PropTypes.string.isRequired
}

// Stolen from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomInt = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min; //The maximum is exclusive and the minimum is inclusive
}

const vindicators = [
  { group:`Vindicator`, value: `supernova`, label: `Supernova`},
  { group:`Vindicator`, value: `noob-noob`, label: `Noob-Noob` },
  { group:`Vindicator`, value: `vance_maximus`, label: `Vance Maximus`},
  { group:`Vindicator`, value: `alan_rails`, label: `Alan Rails`},
  { group:`Vindicator`, value: `crocubot`, label: `Crocubot` },
  { group:`Vindicator`, value: `million_ants`, label: `Million Ants`},
  { group:`Vindicator`, value: `morty_smith`, label: `Morty Smith`},
  { group:`Vindicator`, value: `rick_sanchez`, label: `Rick Sanchez`},
  { group:`Vindicator`, value: `lady_katana`, label: `Lady Katana`},
  { group:`Vindicator`, value: `calypso`, label: `Calypso`},
  { group:`Vindicator`, value: `diablo_verde`, label: `Diablo Verde`}
]

const episodes = [
  {
    element: {
      title: "Raising Gazorpazorp"
    },
    facets: [
      {
        group: "Planet",
        value: "gazorpazorp",
        label: "Gazorpazorp"
      },
      {
        group: "Guest character",
        value: "gwendolyn",
        label: "Gwendolyn"
      },
      {
        group: "Guest character",
        value: "ma-sha",
        label: "Ma-Sha"
      },
      {
        group: "Season",
        value: "1",
        label: "1"
      }
    ]
  },
  {
    element: {
      title: "The wedding squanchers"
    },
    facets: [
      {
        group: "Planet",
        value: "squanch",
        label: "Squanch"
      },
      {
        group: "Guest character",
        value: "birdperson",
        label: "Birdperson"
      },
      {
        group: "Guest character",
        value: "squanchy",
        label: "Squanchy"
      },
      {
        group: "Season",
        value: "2",
        label: "2"
      }
    ]
  },
  {
    element: {
      title: "The Rickshank redemption"
    },
    facets: [
      {
        group: "Guest character",
        value: "birdperson",
        label: "Birdperson"
      },
      {
        group: "Planet",
        value: "buttworld",
        label: "Buttworld"
      },
      {
        group: "Season",
        value: "3",
        label: "3"
      }
    ]
  },
  {
    element: {
      title: "Ricksy business"
    },
    facets: [
      {
        group: "Guest character",
        value: "squanchy",
        label: "Squanchy"
      },
      {
        group: "Guest character",
        value: "abradolf_lincler",
        label: "Abradolf Lincler"
      },
      {
        group: "Season",
        value: "1",
        label: "1"
      }
    ]
  },
  {
    element: {
      title: "Close Rick-counters of the Rick kind"
    },
    facets: [
      {
        group: "Guest character",
        value: "ricktiminus_sancheziminius",
        label: "Ricktiminus Sancheziminius"
      },
      {
        group: "Guest character",
        value: "abradolf_lincler",
        label: "Abradolf Lincler"
      },
      {
        group: "Planet",
        value: "buttworld",
        label: "Buttworld"
      },
      {
        group: "Season",
        value: "1",
        label: "1"
      }
    ]
  }
]



export {getRandomInt, vindicators, episodes, EpisodeCard}
