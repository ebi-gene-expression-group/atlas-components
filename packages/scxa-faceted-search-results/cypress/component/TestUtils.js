import episodes from '../fixtures/episodes.json'
import PropTypes from "prop-types";
import styled from "styled-components";
import React from "react";

const facets = () => {
  return episodes.reduce((acc, episode) => acc.concat(episode.facets), [])
}

const getUniqueFacets = () => {
  const allFacets = facets()
  return allFacets
    .filter((facet, index) => allFacets.findIndex((thatFacet) => facet.value === thatFacet.value) === index)
    .map((facet) => ({
      ...facet,
      disabled: false
    }))
}

const getFacetTooltip = () => {
  return Cypress._.shuffle(getUniqueFacets()).find(facet => facet.description)
}

const getFacetWithoutTooltip = () => {
  return Cypress._.shuffle(getUniqueFacets()).find(facet => !facet.description)
}

const getPropsWithTooltip = (facetTooltip) => {
  return {
    facetGroupName: facetTooltip.group,
      facetGroupNameDescription: facetTooltip.description,
    facets: getUniqueFacets().filter(facet => facet.group === facetTooltip.group),
    onChange: () => {}
  }
}

const getPropsWithoutTooltip = () => {
  const facetWithoutTooltip = getFacetWithoutTooltip()
  return {
    facetGroupName: facetWithoutTooltip.group,
      facets: getUniqueFacets().filter(facet => facet.group === facetWithoutTooltip.group),
    onChange: () => {}
  }
}

const ExperimentTableCard = ({title}) =>
  <div>
    <p>{title}</p>
  </div>

ExperimentTableCard.propTypes = {
  title: PropTypes.string.isRequired
}

const IconDiv = styled.div`
  width: 15%;
  text-align: center;
`
IconDiv.displayName = 'IconDiv'

const BitDiv = styled.div`
  width: 25%;
  text-align: center;
`
const DoveDiv = styled.div`
  width: 25%;
  text-align: center;
`

const ExperimentTableHeader = ({onClick}) =>
  <DoveDiv>
    <IconDiv onClick={()=>onClick(`test`)}>
      Meow
    </IconDiv>
    <BitDiv onClick={()=>onClick(`test`)}>
      Wow wow
    </BitDiv>
  </DoveDiv>

ExperimentTableHeader.propTypes = {
  onClick: PropTypes.func.isRequired
}

// Stolen from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomInt = (min, max) => {
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(Math.random() * (max - min)) + min //The maximum is exclusive and the minimum is inclusive
}

export { episodes, getRandomInt, getUniqueFacets, getFacetTooltip, getPropsWithTooltip, getPropsWithoutTooltip,
  ExperimentTableCard, ExperimentTableHeader }
