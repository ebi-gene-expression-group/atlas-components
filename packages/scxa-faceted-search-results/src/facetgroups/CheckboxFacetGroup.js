import React from 'react'
import PropTypes from 'prop-types'
import { xorBy as _xorBy } from 'lodash'

import FacetGroupPropTypes from './FacetGroupPropTypes'
import TooltipIcon from './TooltipIcon'

const CheckboxOption = ({ group, value, label, disabled, checked, onChange }) =>
  <div>
    <input type={`checkbox`} {...{value, checked, disabled}}
      onChange={() => onChange({group, label, value, disabled})}/>
    <label style={disabled ? {color: `lightgrey`} : {}}>{label}</label>
  </div>

CheckboxOption.propTypes = {
  group: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired
}

// In principle we donâ€™t need this component to be stateful, but in doing so we can create a custom _handleChange
// function that will ultimately call onChange(facetGroupName, facets); this allows us to have the same API as
// React-Select and reuse the same callback for both checkbox-style and multiselect facet groups
class CheckboxFacetGroup extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      checkedFacets: []
    }
    this._handleChange = this._handleChange.bind(this)
  }

  // disabled is passed in but not used (it makes tests easier)
  _handleChange(facet) {
    this.setState(
      { checkedFacets: _xorBy(this.state.checkedFacets, [facet], `value`) },
      () => this.props.onChange(facet.group, this.state.checkedFacets))
  }

  render() {
    const { facetGroupName, facetGroupNameDescription, facets } = this.props
    const { checkedFacets } = this.state

    return (
      <div className={`padding-bottom-xlarge`}>
        <h4>
          { facetGroupName }
          { facetGroupNameDescription && <TooltipIcon tooltipText={facetGroupNameDescription}/> }
        </h4>
        {facets.map((facet) =>
          <CheckboxOption {...facet}
            checked={checkedFacets.some((checkedFacet) => checkedFacet.value === facet.value)}
            onChange={this._handleChange}
            key={facet.value}/>
        )}
      </div>
    )
  }
}

CheckboxFacetGroup.propTypes = FacetGroupPropTypes

export default CheckboxFacetGroup
