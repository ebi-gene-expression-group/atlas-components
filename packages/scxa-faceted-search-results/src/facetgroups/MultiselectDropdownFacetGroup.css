/* Unfortunately we need to set some EBI VF overrides here because it applies styles to elements deper than
React-Selectâ€™s APIs (e.g. the div that encloses the input rather than the input itself) */
.input-clear input, .input-clear input:focus {
  height: unset;
  box-shadow: none;
  margin: 0;
}
