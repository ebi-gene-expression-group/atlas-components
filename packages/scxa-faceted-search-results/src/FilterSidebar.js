import React from 'react'
import PropTypes from 'prop-types'
import _ from 'lodash'

import FacetGroupPropTypes from './facetgroups/FacetGroupPropTypes'
import CheckboxFacetGroup from './facetgroups/CheckboxFacetGroup'
import MultiselectDropdownFacetGroup from './facetgroups/MultiselectDropdownFacetGroup'

import {FacetPropTypes} from './ResultPropTypes'

const FilterSidebar = ({facets, checkboxFacetGroups, onChange}) => {
  const facetGroups =
    _(facets)
      .sortBy([`group`, `label`])
      .groupBy(`group`)
      .toPairs()
      .partition((facetGroup) => checkboxFacetGroups.includes(facetGroup[0]))
      .value()

  // Facets as checkboxes go first by design
  return(
    [
      facetGroups[0]
        .map((facetGroup) => <CheckboxFacetGroup facetGroupName={facetGroup[0]}
                                                 facets={facetGroup[1]}
                                                 onChange={onChange}
                                                 key={facetGroup[0]} />),
      facetGroups[1]
        .map((facetGroup) => <MultiselectDropdownFacetGroup facetGroupName={facetGroup[0]}
                                                            facets={facetGroup[1]}
                                                            onChange={onChange}
                                                            key={facetGroup[0]} />)
    ]
  )
}

FilterSidebar.propTypes = {
  facets: PropTypes.arrayOf(PropTypes.shape({
    ...FacetPropTypes,
    disabled: PropTypes.bool.isRequired})).isRequired,
  checkboxFacetGroups: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func.isRequired
}

FilterSidebar.defaultProps = {
  checkboxFacetGroups: []
}

export default FilterSidebar
