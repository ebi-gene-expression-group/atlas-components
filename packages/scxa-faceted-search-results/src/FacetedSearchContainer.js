import React from 'react'
import ReactTooltip from 'react-tooltip'
import PropTypes from 'prop-types'
import _ from 'lodash'

import FilterSidebar from './FilterSidebar'
import FilterList from './FilterList'

import {ResultPropTypes} from './ResultPropTypes'

class FacetedSearchContainer extends React.Component {
  constructor(props) {
    super(props)

    const uniqueFacets = _(props.results)
      .flatMap(`facets`)
      .compact()   // lodash will emit undefined if `facets` doesn’t exist :/
      .uniqWith(_.isEqual)
      .map((facet) => ({...facet, disabled: false}))
      .value()

    const groupedResultFacets = props.results.map(result => _.groupBy(result.facets, `group`))
    const groupedUniqueFacets = Object.keys(_.groupBy(uniqueFacets, `group`))

    const sharedFacets =
      groupedUniqueFacets.filter(group =>
        !groupedResultFacets.every((result, idx, results) =>
          _.isEqual(result[group], results[0][group])))

    this.state = {
      facets: uniqueFacets.filter(facet => sharedFacets.includes(facet.group)),
      selectedFacets: {} // TODO (?) Build initial state of checked filters from props if wrapped in React Router,
    }

    this._handleChange = this._handleChange.bind(this)
    this._filterResults = this._filterResults.bind(this)
  }

  _filterResults(facets) {
    return this.props.results.filter((result) =>
      // Leave the results that, for every group, have at least one matching property within that group
      // Note: every returns true for empty arrays
      Object.entries(facets)
        .every(
          ([facetGroup, facetValues]) =>
            result.facets.some(
              (resultFacet) =>
                facetGroup === resultFacet.group && facetValues.map(f => f.value).includes(resultFacet.value))))
  }

  _hasNoResults(selectedFacets, additionalFacet) {
    const mergedFacets =
      _.mergeWith(
        _.defaultsDeep({}, selectedFacets),
        {[additionalFacet.group]: additionalFacet},
        (objValue, srcValue) => _.uniq((objValue || []).concat(srcValue))
      )

    return this._filterResults(mergedFacets).length === 0
  }

  _disableEnabledFacetsWithNoResults(selectedFacets, facetGroup) {
    return _.cloneDeep(this.state.facets)
      .map((facet) => ({
        ...facet,
        disabled:
          facet.group !== facetGroup ?
            facet.disabled ? true : this._hasNoResults(selectedFacets, facet) :
            facet.disabled
      }))
  }

  _enableDisabledFacetsWithResults(selectedFacets, facetGroup) {
    return _.cloneDeep(this.state.facets)
      .map((facet) => ({
        ...facet,
        disabled:
          facet.group !== facetGroup ?
            facet.disabled ? this._hasNoResults(selectedFacets, facet) : false :
            facet.disabled
      }))
  }

  _handleChange(facetGroup, selectedFacetsInGroup) {
    const _selectedFacets = _.defaultsDeep({}, this.state.selectedFacets)
    _selectedFacets[facetGroup] = selectedFacetsInGroup

    const nextSelectedFacets =
       Object.keys(_selectedFacets)
         .filter((key) => _selectedFacets[key] && _selectedFacets[key].length > 0)
         .reduce((acc, key) => {
           acc[key] = _selectedFacets[key]
           return acc
         }, {})

    const previousNumberOfSelectedFacetsInGroup = this.state.selectedFacets[facetGroup] ?
      this.state.selectedFacets[facetGroup].length : 0

    let nextFacets = {}
    if (_selectedFacets[facetGroup].length > previousNumberOfSelectedFacetsInGroup) {
      if (_selectedFacets[facetGroup].length === 1) {
        // First facet in group selected: less results, disable enabled facets
        nextFacets = this._disableEnabledFacetsWithNoResults(nextSelectedFacets, facetGroup)
      } else {
        // Add a second or subsequent facet to a group: more results, enable disabled facets
        nextFacets = this._enableDisabledFacetsWithResults(nextSelectedFacets, facetGroup)
      }
    } else {
      if (_selectedFacets[facetGroup].length === 0) {
        // No facets in group selected: more results, enable disabled facets
        nextFacets = this._enableDisabledFacetsWithResults(nextSelectedFacets, facetGroup)
      } else {
        // Facet has been deselected but others in group remain: less results, disable enabled facets
        nextFacets = this._disableEnabledFacetsWithNoResults(nextSelectedFacets, facetGroup)
      }
    }

    this.setState({
      facets: nextFacets,
      selectedFacets: nextSelectedFacets
    })
  }

  render() {
    const {facets, selectedFacets} = this.state
    const {checkboxFacetGroups, ResultElementClass, ResultsHeaderClass, resultsMessage, sortTitle} = this.props

    return(
      <div className={`row expanded`}>
        {
          facets.length > 0 &&
          <div className={`small-12 medium-4 large-3 columns`}>
            <FilterSidebar {...{checkboxFacetGroups, facets}} onChange={this._handleChange}/>
          </div>
        }

        <div className={`small-12 medium-8 large-9 columns`}>
          <FilterList {...{resultsMessage, ResultElementClass, ResultsHeaderClass, sortTitle}}
            filteredResults={this._filterResults(selectedFacets)}/>
        </div>
        <ReactTooltip effect={`solid`}/>
      </div>
    )
  }
}

FacetedSearchContainer.propTypes = {
  // A list of results from where facets can be inferred by inspecting each result’s facets field
  // Supplied by FetchLoader, or set manually for testing
  results: PropTypes.arrayOf(ResultPropTypes).isRequired,
  checkboxFacetGroups: PropTypes.arrayOf(PropTypes.string),
  resultsMessage: PropTypes.string,
  // Must be classes that extends React.Component, sadly there’s no such prop type :(
  // See also https://stackoverflow.com/questions/45315918/react-proptypes-component-class
  ResultsHeaderClass: PropTypes.func,
  ResultElementClass: PropTypes.func.isRequired,
	sortTitle: PropTypes.string
}

FacetedSearchContainer.defaultProps = {
  resultsMessage: ``,
  checkboxFacetGroups: []
}

export default FacetedSearchContainer
