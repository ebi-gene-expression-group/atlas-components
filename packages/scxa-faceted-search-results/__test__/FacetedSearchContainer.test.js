import React from 'react'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import { mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import {episodes, ExperimentTableHeader, ExperimentTableCard} from './TestUtils'

import FacetedSearchContainer from '../src/FacetedSearchContainer'
import FilterSidebar from '../src/FilterSidebar'
import CheckboxFacetGroup from '../src/facetgroups/CheckboxFacetGroup'
import MultiselectDropdownFacetGroup from '../src/facetgroups/MultiselectDropdownFacetGroup'

Enzyme.configure({ adapter: new Adapter() })

const props = {
  results: episodes,
  checkboxFacetGroups: [`Season`],
  ResultsHeaderClass: ExperimentTableHeader,
  ResultElementClass: ExperimentTableCard
}

describe(`FacetedSearchContainer`, () => {
  test(`when results have no facets only the results list is displayed`, () => {
    const wrapperWithFacets = mount(<FacetedSearchContainer {...props} />)
    expect(wrapperWithFacets.find(FilterSidebar)).toHaveLength(1)
    const wrapperWithoutFacets = mount(<FacetedSearchContainer {...props} results={props.results.map((result) => ({element: result.element}))}/>)
    expect(wrapperWithoutFacets.find(FilterSidebar)).toHaveLength(0)
  })

  test(`clicking to select/unselect a single facet in a group works`, () => {
    const wrapper = mount(<FacetedSearchContainer {...props} />)
    wrapper.find(FilterSidebar).find({ type: `checkbox` }).first().simulate(`change`)
    expect(wrapper.find(ExperimentTableCard).length).toBeLessThan(props.results.length)
    wrapper.find(FilterSidebar).find({ type: `checkbox` }).first().simulate(`change`)
    expect(wrapper.find(ExperimentTableCard).length).toBe(props.results.length)
  })

  test(`clicking on a second facet works`, () => {
    const wrapper = mount(<FacetedSearchContainer {...props} />)
    wrapper.find(FilterSidebar).find({ type: `checkbox` }).first().simulate(`change`)
    const numberOfResultsAfterFirstClick = wrapper.find(ExperimentTableCard).length
    wrapper.find(FilterSidebar).find({ type: `checkbox` }).at(1).simulate(`change`)
    expect(wrapper.find(ExperimentTableCard).length).toBeGreaterThan(numberOfResultsAfterFirstClick)
  })

  test(`doesnâ€™t display duplicated facets`, () => {
    const allFacets = episodes.reduce((acc, episode) => acc.concat(episode.facets), [])
    const uniqueFacetsByGroup =
      allFacets
        .filter((facet, index) => allFacets.findIndex((thatFacet) => facet.value === thatFacet.value) === index)
        .reduce((acc, facet) => {
          acc[facet.group] = (acc[facet.group] || []).concat(facet.value)
          return acc
        }, {})

    const wrapper = mount(<FacetedSearchContainer {...props} />)

    wrapper.find(CheckboxFacetGroup).forEach(
      (facetGroup) =>
        expect(facetGroup.props().facets).toHaveLength(uniqueFacetsByGroup[facetGroup.props().facetGroupName].length))

    wrapper.find(MultiselectDropdownFacetGroup).forEach(
      (facetGroup) =>
        expect(facetGroup.props().facets).toHaveLength(uniqueFacetsByGroup[facetGroup.props().facetGroupName].length))
  })

  test(`matches snapshot`, () => {
    const tree = renderer.create(<FacetedSearchContainer {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
