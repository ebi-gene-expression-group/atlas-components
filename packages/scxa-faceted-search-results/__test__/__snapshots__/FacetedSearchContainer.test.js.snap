// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacetedSearchContainer matches snapshot 1`] = `
<div
  className="row expanded"
>
  <div
    className="small-12 medium-4 large-3 columns"
  >
    <div
      className="padding-bottom-xlarge"
    >
      <h4>
        Season
      </h4>
      <div>
        <input
          checked={false}
          disabled={false}
          onChange={[Function]}
          type="checkbox"
          value="1"
        />
        <label
          style={Object {}}
        >
          Season 1
        </label>
      </div>
      <div>
        <input
          checked={false}
          disabled={false}
          onChange={[Function]}
          type="checkbox"
          value="2"
        />
        <label
          style={Object {}}
        >
          Season 2
        </label>
      </div>
      <div>
        <input
          checked={false}
          disabled={false}
          onChange={[Function]}
          type="checkbox"
          value="3"
        />
        <label
          style={Object {}}
        >
          Season 3
        </label>
      </div>
    </div>
    <div
      className="padding-bottom-xlarge"
    >
      <h4>
        Guest character
      </h4>
      <div
        className="css-10nd86i"
        onKeyDown={[Function]}
      >
        <div
          className="css-12emboz"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="css-1hwfws3"
          >
            <div
              className="css-1492t68"
            >
              Select...
            </div>
            <div
              className="css-1g6gooi"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-10-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="css-1wy0on6"
          >
            <span
              className="sc-bdVaJa jcfyVD"
              cx={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="padding-bottom-xlarge"
    >
      <h4>
        Planet
      </h4>
      <div
        className="css-10nd86i"
        onKeyDown={[Function]}
      >
        <div
          className="css-12emboz"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="css-1hwfws3"
          >
            <div
              className="css-1492t68"
            >
              Select...
            </div>
            <div
              className="css-1g6gooi"
            >
              <div
                className=""
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-11-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="css-1wy0on6"
          >
            <span
              className="sc-bdVaJa jcfyVD"
              cx={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="small-12 medium-8 large-9 columns"
  >
    <div>
      <h4>
        
      </h4>
      <div>
        <div>
          <p>
            Raising Gazorpazorp
          </p>
        </div>
      </div>
      <div>
        <div>
          <p>
            The wedding squanchers
          </p>
        </div>
      </div>
      <div>
        <div>
          <p>
            The Rickshank redemption
          </p>
        </div>
      </div>
      <div>
        <div>
          <p>
            Ricksy business
          </p>
        </div>
      </div>
      <div>
        <div>
          <p>
            Close Rick-counters of the Rick kind
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
